src/lib/operators/insert.cpp

```diff
@@ -8,94 +8,80 @@
-    DebugAssert(value_segment, "Cannot insert into non-ValueColumns");
-    auto& values = value_segment->values();
+void resize_value_segment(std::shared_ptr<BaseSegment> segment, size_t new_size) {
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(segment);
+  Assert(value_segment, "Cannot insert into non-ValueColumns");
 
-    values.resize(new_size);
+  value_segment->values().resize(new_size);
 
-    if (value_segment->is_nullable()) {
-      value_segment->null_values().resize(new_size);
-    }
+  if (value_segment->is_nullable()) {
+    value_segment->null_values().resize(new_size);
   }
+}
 
-  // this copies
-  void copy_data(std::shared_ptr<const BaseSegment> source, ChunkOffset source_start_index,
-                 std::shared_ptr<BaseSegment> target, ChunkOffset target_start_index, ChunkOffset length) override {
-    auto casted_target = std::dynamic_pointer_cast<ValueSegment<T>>(target);
-    DebugAssert(casted_target, "Cannot insert into non-ValueColumns");
-    auto& values = casted_target->values();
-
-    auto target_is_nullable = casted_target->is_nullable();
-
-    if (auto casted_source = std::dynamic_pointer_cast<const ValueSegment<T>>(source)) {
-      std::copy_n(casted_source->values().begin() + source_start_index, length, values.begin() + target_start_index);
+template <typename T>
+void copy(const std::shared_ptr<const BaseSegment>& source_base_segment, ChunkOffset source_begin_offset,
+          const std::shared_ptr<BaseSegment>& target_base_segment, ChunkOffset target_begin_offset,
+          ChunkOffset length) {
+  const auto target_value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(target_base_segment);
+  Assert(target_value_segment, "Cannot insert into non-ValueSegments");
+
+  auto& target_values = target_value_segment->values();
+  const auto target_is_nullable = target_value_segment->is_nullable();
+
+  /**
+   * If the source Segment is a ValueSegment, take a fast path to copy the data.
+   * Otherwise, take a (potentially slower) fallback path.
+   */
+  if (const auto source_value_segment = std::dynamic_pointer_cast<const ValueSegment<T>>(source_base_segment)) {
+    std::copy_n(source_value_segment->values().begin() + source_begin_offset, length,
+                target_values.begin() + target_begin_offset);
+
+    if (source_value_segment->is_nullable()) {
+      const auto nulls_begin_iter = source_value_segment->null_values().begin() + source_begin_offset;
+      const auto nulls_end_iter = nulls_begin_iter + length;
+
+      Assert(
+          target_is_nullable || std::none_of(nulls_begin_iter, nulls_end_iter, [](const auto& null) { return null; }),
+          "Trying to insert NULL into non-NULL segment");
+      std::copy(nulls_begin_iter, nulls_end_iter, target_value_segment->null_values().begin() + target_begin_offset);
+    }
+  } else {
+    segment_with_iterators<T>(*source_base_segment, [&](auto source_begin, const auto source_end) {
```

Comment:
Wouldn't `source_it` be more consistent?

Reply:
Nah, why? `segment_with_iterators` gives us the entire Segment's range, but the function just wants to write to a sub-range of it.

Reply:
Aaah, I see what you're doing. Const it?

Reply:
done